compiler/resolution/wrappers.cpp

```diff
@@ -380,27 +407,73 @@ buildDefaultWrapper(FnSymbol* fn,
             wrapper->insertAtTail(new CallExpr(PRIM_MOVE, temp, new CallExpr(PRIM_INIT, new SymExpr(formal->type->symbol))));
         }
       } else {
+        // use argument default for the formal argument
         BlockStmt* defaultExpr = formal->defaultExpr->copy();
         for_alist(expr, defaultExpr->body) {
           wrapper->insertAtTail(expr->remove());
         }
-        if (formal->intent != INTENT_INOUT) {
-          wrapper->insertAtTail(new CallExpr(PRIM_MOVE, temp, wrapper->body->body.tail->remove()));
+
+        // Normally, addLocalCopiesAndWritebacks will handle
+        // adding the copies. However, because of some issues with
+        // default constructors, the copy is added here for them.
+        // (In particular, the called constructor function does not
+        //  include the necessary copies, because it would interfere
+        //  with the array-domain link in
+        //    record { var D={1..2}; var A:[D] int }
+        //  )
+        if (specializeDefaultConstructor) {
+          // Copy construct from the default value.
+          // Sometimes, normalize has already added an initCopy in the
+          // defaultExpr. But if it didn't, we need to add a copy.
+          Expr* fromExpr = wrapper->body->body.tail->remove();
+          bool needsInitCopy = true;
+          if (CallExpr* fromCall = toCallExpr(fromExpr)) {
+            Expr* base = fromCall->baseExpr;
+            if (UnresolvedSymExpr* urse = toUnresolvedSymExpr(base)) {
+              if (0 == strcmp(urse->unresolved, "chpl__initCopy") ||
+                  0 == strcmp(urse->unresolved, "_createFieldDefault"))
+                needsInitCopy = false;
+            } else {
+              INT_ASSERT(0); // if resolved, check for FLAG_INIT_COPY_FN
```

Comment:
I'm not that confident in the diagnosis. The note about FLAG_INIT_COPY_FN is just there b/c it wasn't necessarily set (which is why we're strcmp'ing). That's a possible cause for a failure but I don't think it's necessarily the likely reason someone would run into this assert.